internal/core/core.go

```diff
@@ -0,0 +1,272 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Package core provides a validator for being able to
+// check for core files on DUT's before and after test
+// modules runs.
+package core
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"regexp"
+	"sync"
+	"text/template"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/openconfig/ondatra"
+	"github.com/openconfig/ondatra/binding"
+	"github.com/openconfig/ondatra/eventlis"
+	"google.golang.org/grpc"
+
+	fpb "github.com/openconfig/gnoi/file"
+	opb "github.com/openconfig/ondatra/proto"
+)
+
+var (
+	vendorCoreFilePath = map[opb.Device_Vendor]string{
+		opb.Device_JUNIPER: "/var/core/",
+		opb.Device_CISCO:   "/misc/disk1/",
+		opb.Device_NOKIA:   "/var/core/",
+		opb.Device_ARISTA:  "/var/core/",
+	}
+	vendorCoreFileNamePattern = map[opb.Device_Vendor]*regexp.Regexp{
+		opb.Device_JUNIPER: regexp.MustCompile(".*.tar.gz"),
+		opb.Device_CISCO:   regexp.MustCompile("/misc/disk1/.*core.*"),
+		opb.Device_NOKIA:   regexp.MustCompile("/var/core/coredump-.*"),
+		opb.Device_ARISTA:  regexp.MustCompile("/var/core/core.*"),
+	}
+)
+
+var (
+	validator validatorImpl
+)
+
+type fileInfo struct {
+	Name     string
+	Path     string
+	Modified uint64
+	Existing bool
+}
+
+type dutCoreFiles struct {
+	DUT    string
+	Files  coreFiles
+	Status string
+}
+
+type coreFiles map[string]fileInfo
+
+type checker struct {
+	dut        binding.DUT
+	fileClient fpb.FileClient
+
+	mu        sync.Mutex
+	startTime time.Time
+	endTime   time.Time
+	initial   coreFiles
+	delta     coreFiles
+}
+
+func newChecker(dut binding.DUT) (*checker, error) {
+	dutVendor := dut.Vendor()
+	// vendorCoreFilePath and vendorCoreProcName should be provided to fetch core file on dut.
+	if _, ok := vendorCoreFilePath[dutVendor]; !ok {
+		return nil, fmt.Errorf("add support for vendor %v in var vendorCoreFilePath", dutVendor)
+	}
+	if _, ok := vendorCoreFileNamePattern[dutVendor]; !ok {
+		return nil, fmt.Errorf("add support for vendor %v in var vendorCoreFileNamePattern", dutVendor)
+	}
+	gClients, err := dut.DialGNOI(context.Background(), grpc.WithBlock())
+	if err != nil {
+		return nil, err
+	}
+	return &checker{
+		dut:        dut,
+		fileClient: gClients.File(),
+		initial:    coreFiles{},
+		delta:      coreFiles{},
+		startTime:  time.Now(),
+	}, nil
+}
+
+func (c *checker) check() (coreFiles, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	cores, err := c.checkCores()
+	if err != nil {
+		return nil, err
+	}
+	if len(c.initial) == 0 {
+		for k, v := range cores {
+			v.Existing = true
+			c.initial[k] = v
+		}
+		return c.initial, nil
+	}
+	for k, v := range cores {
+		if _, ok := c.initial[k]; !ok {
+			c.delta[k] = v
```

Comment:
this delta is always the delta from the initial as opposed to the last time check() was called. I'd probably replace initial and delta fields with only a `prevCores` field, i.e. `newChecker` returns

```
&checker{
	dut:        dut,
	fileClient: gClients.File(),
	prevCores:  coreFiles{},
	startTime:  time.Now(),
}, nil
```

and then the if statement and for loop here can be replaced by:

```
delta := coreFiles{}
for k, v := range cores {
	if _, ok := c.prevCores[k]; !ok {
		delta[k] = v
	}
}
c.prevCores = cores
return delta, nil